### PR TITLE
Update the Consensus Game

### PR DESCRIPTION
**Pull request by geoknee**
To reduce ambiguity in how a consensus game should work, I have:

- [x] Updated the top level readme of this repo
- [x] Added an `updateType` field to the application attributes
- [x] Drawn a mermaid diagram for the valid transitions, which each have a name
- [x] Wrote a short readme with some examples
- [x] overhauled the `validTransition` function accordingly
- [x] modified a single test to pass under the new contracts

and plan on 
- [ ] finishing off comprehensive testing
- [ ] considering using `SafeMath` to sum total funds allocated
- [ ] make helper functions for each (all?) named transitions: such as `accept`, `reject`, which take the previous players commitment as input and craft the appropriate commitment to transition to

Screenshot 2019-05-02 at 15 10 12

### CODE REVIEW

File: packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol [hunk 1]
```diff
@@ -13,62 +13,120 @@ contract ConsensusApp {
...
+        _;
+    }
+
+    modifier proposalsUnchanged(ConsensusCommitment.ConsensusCommitmentStruct memory oldCommitment, ConsensusCommitment.ConsensusCommitmentStruct memory newCommitment) {
+      require(encodeAndHashAllocation(oldCommitment.proposedAllocation) == encodeAndHashAllocation(newCommitment.proposedAllocation),"ConsensusApp:  'proposedAllocation' must be the same between commitments."); 
+      require(encodeAndHashDestination(oldCommitment.proposedDestination) == encodeAndHashDestination(newCommitment.proposedDestination),"ConsensusApp:  'proposedDestination' must be the same between commitments."); 
+        _;
+    }
+
+    modifier totalAllocationConserved(ConsensusCommitment.ConsensusCommitmentStruct memory oldCommitment, ConsensusCommitment.ConsensusCommitmentStruct memory newCommitment) {
```
Comment (geoknee) [inline packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol]: Is this something we want to enforce?

**Comment by geoknee** [comment]
Very open to improvements to names. A suggested change set:

Accord -> Consensus
Motion -> Proposal
Modify -> Repropose

File: packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol (hunk 1 above)
Comment (tomclose) [reply packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol]: No - it makes top-ups impossible.

Comment (geoknee) [reply packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol]: Good point!

File: packages/fmg-nitro-adjudicator/contracts/ConsensusCommitment.sol [hunk 2]
```diff
@@ -1,38 +1,43 @@
...
 import "fmg-core/contracts/Commitment.sol";
 
 library ConsensusCommitment {
     using Commitment for Commitment.CommitmentStruct;
+    
+    enum UpdateType { Accord, Motion }
 
     struct AppAttributes {
-        uint32 consensusCounter;
+        uint32 numVotes;
```
Comment (tomclose) [inline packages/fmg-nitro-adjudicator/contracts/ConsensusCommitment.sol]: Maybe change this to `furtherVotesNeeded`?

File: packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol [hunk 3]
```diff
@@ -13,62 +13,120 @@ contract ConsensusApp {
...
+    currentsUnchanged(oldCommitment, newCommitment)
+    proposalsUnchanged(oldCommitment, newCommitment)
+    { }
 
-        revert('ConsensusApp: Invalid input -- consensus counters out of range');
+    function validatePropose(ConsensusCommitment.ConsensusCommitmentStruct memory oldCommitment, ConsensusCommitment.ConsensusCommitmentStruct memory newCommitment)
+    private
+    pure
+    currentsUnchanged(oldCommitment, newCommitment)
+    totalAllocationConserved(oldCommitment, newCommitment)
```
Comment (tomclose) [inline packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol]: Need to check the `numVotes` is set appropriately.

Comment (tomclose) [reply packages/fmg-nitro-adjudicator/contracts/ConsensusApp.sol]: Oh, maybe this is done above.

**Comment by geoknee** [comment]
I was wondering about the values of `proposedAllocation` and `proposedDestination` for Accord (soon to be Consensus) commitments. 

1. How about they just equal their `allocation` and `destination` counterparts? 
2. Do we actually need to check anything about the proposal fields on Consensus ever? As long as those commitments are identified as such by `updateType`, the proposal fields can be safely ignored, can't they?

**Comment by lalexgap** [comment]
Going to close this since #64 includes this work.

**Pull request closed without merging**